ruby_programming/advanced_ruby/blocks.md

```diff
@@ -0,0 +1,547 @@
+a_proc.call("tim", 80)
+~~~
+
+There isn't much more in creating Procs that you haven't seen with Lambdas. So why use one over the other? Well there are some important differences between them which we'll cover next.
+
+### Procs vs Lambdas
+
+There are some key differences between Procs and Lambdas that can make choosing one over the other more suitable.
+
+#### Arguments
+
+A Proc behaves much like a block with regards to arguments. A proc doesn't care if you give it none, less or more arguments than you specify. It will assign nil to any parameters you name but don't pass through as arguments.
+
+~~~ruby
+a_proc = Proc.new { |a, b| puts "a: #{a} --- b: #{b}" }
+
+a_proc.call("apple")
+
+# => a: apple --- b:
+~~~
+
+A lambda, on the other hand, does care and will raise an error if you don't honor the number of parameters expected
+
+~~~ruby
+a_proc = lambda { |a, b| puts "a: #{a} --- b: #{b}" }
+
+a_proc.call("apple")
+
+# => wrong number of Arguments (given 1, expected 2) (ArgumentError)
+
+a_proc.call("apple", "banana", "cake")
+
+# => wrong number of Arguments (given 3, expected 2) (ArgumentError)
+~~~
+
+#### Returning
+
+When you write an explicit return inside a lambda it returns from the lambda block back to the caller
+
+~~~ruby
+a_lambda = -> { return 1 }
+a_lambda.call
+
+# => 1
+~~~
+
+A Proc object returns from the context in which it is called. If you are in the top level context (outside of a class and method) then you'll get an error because you can't return out of the very top level context, as there is no caller to return to.
+
+~~~ruby
+a_proc = Proc.new { return }
+
+a_proc.call
+
+# => localJumpError (unexpected return)
+~~~
+
+If you try the above in on repl.it you won't get an error, which is just to do with how they manage the context of code. If you try it in irb then you'd get the expected error.
+
+If you return from a proc inside a method, the method is the context in which it was called and therefore it returns from the method before any of the other code below it is executed.
+
+~~~ruby
+def my_method
+  a_proc = Proc.new { return }
+  a_proc.call
+  puts "this line is never reached"
+end
+
+my_method
+~~~
+
+### Similarities
+
+Although we've covered some big differences between procs and lambdas there are some similarities we need to cover.
+
+#### Default Arguments
+
+Both procs and lambdas support default arguments in the same way Ruby methods do
+
+~~~ruby
+my_proc = Proc.new { |name="bob"| puts name }
+
+my_proc.call
+
+# => bob
+
+my_lambda = ->(name="r2d2") { puts name }
+
+my_lambda.call
+
+# => r2d2
+~~~
+
+#### Method parameters
+
+Both procs and lambdas can be used as arguments to a method
+
+~~~ruby
+def my_method(useful_arg)
+  useful_arg.call
+end
+
+my_lambda = -> { puts "lambda" }
+my_proc = Proc.new { puts "proc" }
+
+my_method(my_lambda)
+
+# => lambda
+
+my_method(my_proc)
+
+# => proc
+
+# You can even create the lambda or proc when calling the method
+
+my_method(-> { puts "cool lambda bro" })
```

Comment:
I'm on the fence about including this example, some students might actually use it in their code 😆 

![](https://jrlburke.files.wordpress.com/2015/10/could-should.jpg?w=620)

Reply:
I just removed it. I included it initially because I wanted to show it's nothing special, just an argument like any other but on reflection it's not something that should be done.